src/main/java/com/hubspot/imap/utils/parsers/fetch/EnvelopeParser.java

```diff
@@ -79,8 +112,49 @@ public Envelope parse(List<Object> in) {
     return envelope.build();
   }
 
+  public static Envelope fromHeader(Header header) {
+    Map<String, String> envelopeFields = header.getFields().stream()
+        .filter(f -> EnvelopeField.NAME_INDEX.containsKey(f.getName().toLowerCase()))
+        .collect(Collectors.toMap(f -> f.getName().toLowerCase(), Field::getBody));
+    Envelope.Builder envelope = new Envelope.Builder();
+
+    String dateString = envelopeFields.get(EnvelopeField.DATE.getFieldName());
+    List<ImapAddress> fromAddress = emailAddressesFromStringList(envelopeFields.get(EnvelopeField.FROM.getFieldName()));
+    envelope.setDateString(dateString)
+        .setSubject(envelopeFields.get(EnvelopeField.SUBJECT.getFieldName()))
+        .setFrom(fromAddress)
+        .setSender(emailAddressesFromStringList(envelopeFields.get(EnvelopeField.SENDER.getFieldName()), fromAddress))
+        .setReplyTo(emailAddressesFromStringList(envelopeFields.get(EnvelopeField.REPLY_TO.getFieldName()), fromAddress))
+        .setTo(emailAddressesFromStringList(envelopeFields.get(EnvelopeField.TO.getFieldName())))
+        .setCc(emailAddressesFromStringList(envelopeFields.get(EnvelopeField.CC.getFieldName())))
+        .setBcc(emailAddressesFromStringList(envelopeFields.get(EnvelopeField.BCC.getFieldName())))
+        .setInReplyTo(envelopeFields.get(EnvelopeField.IN_REPLY_TO.getFieldName()))
+        .setMessageId(envelopeFields.get(EnvelopeField.MESSAGE_ID.getFieldName()));
+
+    try {
+      if (!Strings.isNullOrEmpty(dateString)) {
+        envelope.setDate(parseDate(dateString));
+      }
+    } catch (DateTimeParseException e) {
+      LOGGER.debug("Failed to parse date {}", header.getField("date").getBody(), e);
+    }
+    return envelope.build();
+  }
+
+  private static List<ImapAddress> emailAddressesFromStringList(String addresses) {
+    return Strings.isNullOrEmpty(addresses) ?
+        Collections.EMPTY_LIST :
+        COMMA_SPLITTER.splitToList(addresses).stream().map(address -> new Builder().setAddress(address)).collect(Collectors.toList());
+  }
+
+  private static List<ImapAddress> emailAddressesFromStringList(String addresses, List<ImapAddress> defaults) {
+    return Strings.isNullOrEmpty(addresses) ?
+        defaults :
+        COMMA_SPLITTER.splitToList(addresses).stream().map(address -> new Builder().setAddress(address)).collect(Collectors.toList());
```

Comment:
NIT: I'd prefer you don't import the builder and instead here declare `new ImapAddress.Builder()` for clarity